lib/ace/ext/papeeria/papeeria_latex_highlight_rules_test.js

```diff
@@ -0,0 +1,176 @@
+
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_NUMBERED_STATE].end = "\\end{equation}";
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_STATE].end = "\\end{equation*}";
+    mathConstants[MATH_TEX_INLINE_STATE].end = "$";
+    mathConstants[MATH_TEX_DISPLAYED_STATE].end = "$$";
+    mathConstants[MATH_LATEX_INLINE_STATE].end = "\\)";
+    mathConstants[MATH_LATEX_DISPLAYED_STATE].end = "\\]";
+
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_NUMBERED_STATE].length = 4;
+    mathConstants[MATH_ENVIRONMENT_DISPLAYED_STATE].length = 4;
+    mathConstants[MATH_TEX_INLINE_STATE].length = 1;
+    mathConstants[MATH_TEX_DISPLAYED_STATE].length = 1;
+    mathConstants[MATH_LATEX_INLINE_STATE].length = 1;
+    mathConstants[MATH_LATEX_DISPLAYED_STATE].length = 1;
+
+    module.exports = {
+        "test: basic testing": function() {
+            var rules = new PapeeriaLatexHighlightRules().getRules();
+            var tokenizer = new Tokenizer(rules);
+            var basicString = " x^2 \\text{hey} \\alpha \\$ % this is a comment";
+            for (var state in mathConstants) {
+                var stateConstants = mathConstants[state];
+                var testString = stateConstants.start + basicString;
+                var result = tokenizer.getLineTokens(testString);
+                var tokens = result.tokens;
+                var state = result.state;
+                assert.ok(state.length);
+                for (var i = stateConstants.length; i < tokens.length; i++) {
+                    assert(isType(tokens[i], EQUATION_TOKEN_TYPE), state);
+                }
+            }
+        },
+
+        "test: mismatching ends": function() {
+            var rules = new PapeeriaLatexHighlightRules().getRules();
+            var tokenizer = new Tokenizer(rules);
+
+            for (var startState in mathConstants) {
+                var startConstants = mathConstants[startState];
+                for (var endState in mathConstants) {
+                    var endConstants = mathConstants[endState];
+                    if (endConstants.end.startsWith(startConstants.end)) {
+                        continue;
+                    }
+
+                    var testString = startConstants.start + " " + endConstants.end;
+                    var result = tokenizer.getLineTokens(testString);
+                    assert.equal(result.state[result.state.length - 1], startState);
+                }
+            }
+        },
+
+        "test: nested equation starts": function() {
+            var rules = new PapeeriaLatexHighlightRules().getRules();
+            var tokenizer = new Tokenizer(rules);
+
+            for (var initState in mathConstants) {
+                var initStateConstants = mathConstants[initState];
+                for (var innerState in mathConstants) {
+                    var innerStateConstants = mathConstants[innerState];
+                    if (innerStateConstants.start.startsWith(initStateConstants.end)) {
```

Comment:
Did you really mean `initStateConstants.end`, not `initStateConstants.start` ? This will test e.g. 

```
\begin{equation}\begin{equation}
```

which is probably not bad, but I am not sure if you meant to test it.